search-service/src/main/kotlin/com/egm/stellio/search/service/TemporalEntityAttributeService.kt

```diff
@@ -359,22 +360,42 @@ class TemporalEntityAttributeService(
         entitiesIds: List<URI>,
         entitiesQuery: EntitiesQuery
     ): List<TemporalEntityAttribute> {
-        val filterOnAttributes =
-            if (entitiesQuery.attrs.isNotEmpty())
-                " AND " + entitiesQuery.attrs.joinToString(
-                    separator = ",",
-                    prefix = "attribute_name in (",
-                    postfix = ")"
-                ) { "'$it'" }
-            else ""
+        val expandedAttrsList = entitiesQuery.attrs.joinToString(",") { "'$it'" }
+        val datasetIdsList = entitiesQuery.datasetId.joinToString(",") { "'$it'" }
+
+        val datasetIdDefaultValueQuery = " AND ((dataset_id IS NOT NULL AND dataset_id in ($datasetIdsList))" +
+            "OR dataset_id IS NULL)"
+        val filterQuery = when {
+            entitiesQuery.attrs.isNotEmpty() && entitiesQuery.datasetId.isNotEmpty() -> {
+                if (entitiesQuery.datasetId.contains(NGSILD_NONE_TERM)) {
+                    " AND attribute_name IN ($expandedAttrsList) $datasetIdDefaultValueQuery"
+                } else {
+                    " AND attribute_name IN ($expandedAttrsList) AND dataset_id IN ($datasetIdsList)"
+                }
+            }
+
+            entitiesQuery.attrs.isEmpty() && entitiesQuery.datasetId.isNotEmpty() -> {
+                if (entitiesQuery.datasetId.contains(NGSILD_NONE_TERM)) {
+                    "$datasetIdDefaultValueQuery"
+                } else {
+                    " AND dataset_id IN ($datasetIdsList)"
+                }
+            }
+
+            entitiesQuery.attrs.isNotEmpty() && entitiesQuery.datasetId.isEmpty() -> {
+                " AND attribute_name IN ($expandedAttrsList)"
+            }
+
+            else -> ""
+        }
```

Comment:
that is very (too much) complex.

- there was already a filter clause for `attrs`
- for `datasetId`, you create such a filter (or null / empty if no `datasetId` is provided): 

```sql
AND ((dataset_id IS NOT NULL AND dataset_id in ($datasetIdsList)
    OR (dataset_id IS NULL AND '@none' in ($datasetIdsList))
```

- you combine both to create the final filter clause